sdks/python/apache_beam/utils/windowed_value.py

```diff
@@ -249,10 +268,13 @@ class _IntervalWindowBase(object):
   """Optimized form of IntervalWindow storing only microseconds for endpoints.
   """
 
-  def __init__(self, start, end):
+  def __init__(self,
+               start,  # type: Optional[Union[int, float, Timestamp]]
```

Comment:
Hmm... why are these (and below) optional?

Reply:
Yeah, this code is a bit obtuse. 

There is one place in the code that I can see None being passed, in `IntervalWindowCoderImpl`:

```python
  def decode_from_stream(self, in_, nested):
    # type: (create_InputStream, bool) -> IntervalWindow
    if not TYPE_CHECKING:
      global IntervalWindow
      if IntervalWindow is None:
        from apache_beam.transforms.window import IntervalWindow
    typed_value = IntervalWindow(None, None)
    typed_value._end_micros = (
        1000 * self._to_normal_time(in_.read_bigendian_uint64()))
    typed_value._start_micros = (
        typed_value._end_micros - 1000 * in_.read_var_int64())
    return typed_value
```

The problem with the `None` code path is that calling almost any method on the resulting instance will result in `AttributeError` because neither `_start_micros` nor `_end_micros` is created.  You can see in the example above, those values are patched in, which works, but it's pretty hacky.

It seems like this would be more straight-forward:

```python
  def decode_from_stream(self, in_, nested):
    # type: (create_InputStream, bool) -> IntervalWindow
    if not TYPE_CHECKING:
      global IntervalWindow
      if IntervalWindow is None:
        from apache_beam.transforms.window import IntervalWindow
    end_micros = 1000 * self._to_normal_time(in_.read_bigendian_uint64())
    end = Timestamp(0, end_micros)
    start = Timestamp(0, end_micros - 1000 * in_.read_var_int64())
    typed_value = IntervalWindow(start, end)
    return typed_value
```

Reply:
Ah. The goal here is to avoid creating intermediate Python objects... Maybe worth a comment here.